cpp/tests/mtmg/threaded_test.cpp

```diff
@@ -0,0 +1,469 @@
+    instance_manager->reset_threads();
+
+    std::vector<std::tuple<std::vector<vertex_t>, std::vector<result_t>>> computed_pageranks_v;
+    std::mutex computed_pageranks_lock{};
+
+    // Load computed_pageranks from different threads.  We'll use more threads than GPUs here
+    for (int i = 0; i < num_threads; ++i) {
+      running_threads.emplace_back([&instance_manager,
+                                    &graph_view,
+                                    &pageranks,
+                                    &computed_pageranks_lock,
+                                    &computed_pageranks_v,
+                                    &h_src_v,
+                                    &h_dst_v,
+                                    &h_weights_v,
+                                    i,
+                                    num_threads]() {
+        auto thread_handle = instance_manager->get_handle();
+
+        auto number_of_vertices = graph_view.get_pointer(thread_handle)->number_of_vertices();
+
+        std::vector<vertex_t> my_vertex_list;
+        my_vertex_list.reserve((number_of_vertices + num_threads - 1) / num_threads);
+
+        for (int j = i; j < number_of_vertices; j += num_threads) {
+          my_vertex_list.push_back(j);
+        }
+
+        rmm::device_uvector<vertex_t> d_my_vertex_list(my_vertex_list.size(),
+                                                       thread_handle.raft_handle().get_stream());
+        raft::update_device(d_my_vertex_list.data(),
+                            my_vertex_list.data(),
+                            my_vertex_list.size(),
+                            thread_handle.raft_handle().get_stream());
+
+        auto d_my_pageranks = pageranks.gather(
+          thread_handle,
+          raft::device_span<vertex_t const>{d_my_vertex_list.data(), d_my_vertex_list.size()},
+          graph_view);
+
+        std::vector<result_t> my_pageranks(d_my_pageranks.size());
+        raft::update_host(my_pageranks.data(),
+                          d_my_pageranks.data(),
+                          d_my_pageranks.size(),
+                          thread_handle.raft_handle().get_stream());
+
+        {
+          std::lock_guard<std::mutex> lock(computed_pageranks_lock);
+          computed_pageranks_v.push_back(
+            std::make_tuple(std::move(my_vertex_list), std::move(my_pageranks)));
+        }
+      });
+    }
+
+    // Wait for CPU threads to complete
+    std::for_each(running_threads.begin(), running_threads.end(), [](auto& t) { t.join(); });
+    running_threads.resize(0);
+    instance_manager->reset_threads();
+
+    std::map<vertex_t, result_t> computed_pageranks;
+    std::for_each(
+      computed_pageranks_v.begin(), computed_pageranks_v.end(), [&computed_pageranks](auto tuple) {
+        auto [vertex, pagerank] = tuple;
+        std::for_each(thrust::make_zip_iterator(vertex.begin(), pagerank.begin()),
+                      thrust::make_zip_iterator(vertex.end(), pagerank.end()),
```

Comment:
Yes, the `thrust::zip_iterator` works both in GPU and CPU memory.  I use it fairly regularly in host code (mostly for testing purposes), since STL doesn't offer one yet, and bringing in boost is heavier than we need (since we're already using thrust).